examples/erc20_gas/src/main.rs

```diff
@@ -0,0 +1,327 @@
+            .expect("Balance slot not found")
+            .present_value();
+
+        if storage_value < token_amount {
+            panic!("Insufficient balance");
+        }
+
+        token_account.data.storage.insert(
+            balance_slot,
+            EvmStorageSlot::new_changed(storage_value, storage_value.saturating_sub(token_amount)),
+        );
+
+        // Add tokens to treasury
+        let treasury_balance_slot: U256 = keccak256((TREASURY, U256::from(3)).abi_encode()).into();
+        let treasury_balance = token_account
+            .storage
+            .get(&treasury_balance_slot)
+            .expect("Treasury balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            treasury_balance_slot,
+            EvmStorageSlot::new_changed(
+                treasury_balance,
+                treasury_balance.saturating_add(token_amount),
+            ),
+        );
+
+        Ok(())
+    });
+
+    handler.post_execution.reimburse_caller = Arc::new(|ctx, gas| {
+        let caller = ctx.evm.inner.env.tx.common_fields().caller();
+        let gas_price = ctx.evm.inner.env.effective_gas_price();
+        let refund_amount = gas_price * U256::from(gas.remaining() + gas.refunded() as u64);
+
+        if refund_amount.is_zero() {
+            return Ok(());
+        }
+
+        let token_account = ctx
+            .evm
+            .inner
+            .journaled_state
+            .load_account(TOKEN, &mut ctx.evm.inner.db)
+            .map_err(EVMError::Database)?;
+
+        // Return tokens from treasury to caller
+        let treasury_balance_slot: U256 = keccak256((TREASURY, U256::from(3)).abi_encode()).into();
+        let treasury_balance = token_account
+            .storage
+            .get(&treasury_balance_slot)
+            .expect("Treasury balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            treasury_balance_slot,
+            EvmStorageSlot::new_changed(
+                treasury_balance,
+                treasury_balance.saturating_sub(refund_amount),
+            ),
+        );
+
+        let caller_balance_slot: U256 = keccak256((caller, U256::from(3)).abi_encode()).into();
+        let caller_balance = token_account
+            .storage
+            .get(&caller_balance_slot)
+            .expect("Caller balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            caller_balance_slot,
+            EvmStorageSlot::new_changed(
+                caller_balance,
+                caller_balance.saturating_add(refund_amount),
+            ),
+        );
+
+        Ok(())
+    });
+
+    handler.post_execution.reward_beneficiary = Arc::new(|ctx, gas| {
+        let beneficiary = *ctx.evm.env.block.coinbase();
+        let gas_price = ctx.evm.env.effective_gas_price();
+        let reward = gas_price * U256::from(gas.spent() - gas.refunded() as u64);
```

Comment:
It should be `(gas_price-basefee)*U256::from(...)` as block basefee is discarded.